ducktape/utils/util.py

```diff
@@ -19,26 +19,42 @@
 import time
 
 
-def wait_until(condition, timeout_sec, backoff_sec=.1, err_msg=""):
+def wait_until(condition, timeout_sec, backoff_sec=.1, err_msg="", retry_on_exc=False):
     """Block until condition evaluates as true or timeout expires, whichever comes first.
 
     :param condition: callable that returns True if the condition is met, False otherwise
     :param timeout_sec: number of seconds to check the condition for before failing
     :param backoff_sec: number of seconds to back off between each failure to meet the condition before checking again
     :param err_msg: a string or callable returning a string that will be included as the exception message if the
                     condition is not met
+    :oaram retry_on_exc: if True, will retry if condition raises an exception. If condition raised exception on last
+                         iteration, that exception will be raised as a cause of TimeoutError.
+                         If False and condition raises an exception, that exception will be forwarded to the caller
+                         immediately.
+                         Defaults to False (original ducktape behavior).
     :return: silently if condition becomes true within the timeout window, otherwise raise Exception with the given
     error message.
     """
     start = time.time()
     stop = start + timeout_sec
+    last_exception = None
     while time.time() < stop:
-        if condition():
-            return
-        else:
+        try:
+            if condition():
+                return
+            else:
+                # reset last_exception if last iteration didn't raise any exception, but simply returned False
+                last_exception = None
+        except Exception as e:
```

Comment:
I think all the exceptions (hah!) of errors that inherit from BaseException instead of Exception are ok to skip here, but should double check that we don't want to cast an even wider net.

Reply:
Good idea, I don't think it'll hurt to catch a BaseException here, changed